homeassistant/components/guardian/__init__.py

```diff
@@ -62,12 +95,29 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
             api_lock=api_lock,
             valve_controller_uid=entry.data[CONF_UID],
         )
-        initial_fetch_tasks.append(
-            hass.data[DOMAIN][DATA_COORDINATOR][entry.entry_id][api].async_refresh()
+        init_valve_controller_tasks.append(coordinator.async_refresh())
+
+    await asyncio.gather(*init_valve_controller_tasks)
+
+    # Set up an object to evaluate each batch of paired sensor UIDs and add/remove
+    # devices as appropriate:
+    paired_sensor_manager = PairedSensorManager(hass, entry, client, api_lock)
+    await paired_sensor_manager.async_init()
+
+    @callback
+    def async_new_sensor_pair_dump():
+        """Define a callback for when new paired sensor data is received."""
+        async_dispatcher_send(
+            hass, SIGNAL_NEW_SENSOR_PAIR_DUMP.format(entry.data[CONF_UID])
         )
 
-    await asyncio.gather(*initial_fetch_tasks)
+    # When the sensor_pair_dump API completes, send a signal (via the dispatcher) to the
+    # PairedSensorManager so it can process any paired sensor addition/removal:
+    hass.data[DOMAIN][DATA_COORDINATOR][entry.entry_id][
+        API_SENSOR_PAIR_DUMP
+    ].async_add_listener(async_new_sensor_pair_dump)
```

Comment:
Why use a listener to then send a dispatcher? Why not have the thing that's listening to the dispatcher subscribe to the data coordinator itself ?

Reply:
Because the `DataUpdateCoordinator` only accepts sync functions as listeners; I couldn't see a way to add a coroutine (which is what's ultimately needed) as a listener.